File: src/main/javascript/nodyn/bindings/async_wrap.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

function Async(object) {
  this._object = object;
  try {
    process._runAsyncQueue( this );
  } catch (err) {
    this._asyncFlags = this._asyncFlags | 1;
  }

  this._object.on( "makeCallbackByIndex", _makeCallbackByIndex.bind(this) );
}

function _hasAsyncListener() {
  return this._asyncFlags & 1;
}

function _makeCallbackByIndex(result) {
  var index = result.result;
  var callback = this[index];
  _makeCallback.bind(this)( callback );
}

function _makeCallback(callback) {
  if ( _hasAsyncListener.bind(this)() ) {
    try {
      process._loadAsyncQueue( this );
    } catch (err) {
      return;
    }
  }

  var ret;

  try {
    ret = callback.apply( this );
  } catch (err) {
    return;
  }

  if ( _hasAsyncListener.bind(this)() ) {
    try {
      process._unloadAsyncQueue( this );
    } catch(err) {
      return;
    }
  }

  if ( process._tickInfo.inTick ) {
    return ret;
  }
  if ( process._tickInfo[1] === 0 ) {
    process._tickInfo[0] = 0;
    return ret;
  }

  process._tickInfo.inTick = true;

  try {
    process._tickCallback();
  } catch(err) {
    process._tickInfo.lastThrew = true;
  } finally {
    process._tickInfo.inTick = false;
  }

  return ret;

}

module.exports.Async = Async;

File: src/main/java/io/nodyn/async/AsyncWrap.java
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.nodyn.async;

import io.nodyn.CallbackResult;
import io.nodyn.EventSource;
import io.nodyn.NodeProcess;

/**
 * @author Bob McWhirter
 */
public class AsyncWrap extends EventSource {

    protected final NodeProcess process;

    public AsyncWrap(NodeProcess process) {
        this.process = process;
    }

    public NodeProcess getProcess() {
        return this.process;
    }

    public void makeCallback(final int index) {
        this.process.getEventLoop().submitUserTask( new Runnable() {
            @Override
            public void run() {
                emit("makeCallbackByIndex", CallbackResult.createSuccess( index ) );
            }
        }, "make-callback-for-" + getClass().getSimpleName() );

    }

    public Object emit(final String event, final CallbackResult result) {
        this.process.getEventLoop().submitUserTask( new Runnable() {
            @Override
            public void run() {
                AsyncWrap.super.emit( event, result );
            }
        }, "emit-for-" + getClass().getSimpleName() );
        return null;
    }
}
